### ISSUE TITLE
Branchless binary search

### ISSUE THREAD

**Issue by Morwenn** [serge-sans-paille/frozen#2]
Since the size of your `frozen::set` is known at compile time, you should be able to implement a branchless binary search for the lookup when the number of elements is a power of 2 minus 1. I found an old piece of code I wrote some time ago which uses such a branchless binary search:

```cpp
/*
 * This function inserts first in [first+1, first+8), which is a
 * sorted sequence.
 */
template<typename RandomAccessIterator, typename Compare=std::less<>>
auto front_insert8(RandomAccessIterator first, Compare compare={}) const
    -> void
{
    // BEGIN BINARY SEARCH
    auto it = first;
    it += compare(it[4], *first) << 2;
    it += compare(it[2], *first) << 1;
    it += compare(it[1], *first) << 0;
    // END BINARY SEARCH

    switch (std::distance(first, it))
    {
        // rotate_left rotates the sequence beginning at the passed
        // iterator by N positions to the left
    
        case 1: rotate_left<2>(first); break;
        case 2: rotate_left<3>(first); break;
        case 3: rotate_left<4>(first); break;
        case 4: rotate_left<5>(first); break;
        case 5: rotate_left<6>(first); break;
        case 6: rotate_left<7>(first); break;
        case 7: rotate_left<8>(first); break;
        default: break;
    }
}
```

When I benchmarked this version, it was globally faster than the usual one. You could try to specialize your own `lower_bound` function to use a branchless binary search when possible and see whether it's faster :)

Note that it probably isn't better than a regular binary search when the `compare` function it self isn't branchless.

**Comment by serge-sans-paille** [comment]
That's a great trick! i'll benchmark it :-)

**Comment by serge-sans-paille** [comment]
After some tests : the requirements on the size (power of two minus 1) makes it scarcely usefull, and the benfit is not huge, so i'd rather keep current generic version.

**Comment by Morwenn** [comment]
Making it generic shouldn't be too hard with a template dispatch of the `std::bool_constant<is_power_of_two_minus_one(N)>` kind, but if the benefit is not that good, then it might indeed not be worth it. Too bad, it looked fun :/

**Comment by serge-sans-paille** [comment]
Making it generic was indeed not an issue. I just did not report any sensible performance improvement :-/

**Comment by serge-sans-paille** [comment]
For the record, compilers are getting smarter : 
https://godbolt.org/g/zilVYU

**Comment by serge-sans-paille** [comment]
For the record, I implemented a simple branchless version in #25. It still uses a conditional moves but it gives interesting speedup, especially on clang (gcc was already capable of figuring out the ``cmov``

**Comment by Morwenn** [comment]
A branchless trail-recursive implementation? Now that's interesting :p

I'm always surprised by how much compilers are able to optimize away some things to the point that trying to find clever algorithms occasionally becomes pointless ^^'
